demos/device/src/room/containers/RoomProvider/index.tsx

```diff
@@ -0,0 +1,142 @@
+    },
+  };
+  const messageHandler: MessageHandler = async ({ type, payload }) => {
+    console.log(`RoomProvider::messageHandler - Message received with type: ${type}`);
+
+    try {
+      switch (type) {
+        case actionType.JoinMeeting:
+          const { meetingId, name } = payload;
+          if (!meetingId || !name) return;
+
+          await MeetingManager.joinMeeting(meetingId, name);
+          dispatch({ type: actionType.JoinMeeting });
+          history.push(routes.MEETING);
+          MeetingManager.registerScreenShareObservers(screenShareObservers);
+          break;
+        case actionType.StartLocalVideo:
+          MeetingManager.startLocalVideo();
+          dispatch({ type: actionType.StartLocalVideo });
+          break;
+        case actionType.StopLocalVideo:
+          MeetingManager.stopLocalVideo();
+          dispatch({ type: actionType.StopLocalVideo });
+          break;
+        case actionType.LeaveMeeting:
+          await MeetingManager.leaveMeeting();
+          history.push(routes.ROOT);
+          dispatch({ type: actionType.LeaveMeeting });
+          break;
+        case actionType.EndMeeting:
+          await MeetingManager.endMeeting();
+          history.push(routes.ROOT);
+          dispatch({ type: actionType.EndMeeting });
+          break;
+        case actionType.StartScreenShareView:
+          screenViewDiv().style.display = 'grid';
+          MeetingManager.startViewingScreenShare(screenViewDiv());
+          dispatch({ type: actionType.StartScreenShareView });
+          break;
+        case actionType.StopScreenShareView:
+          screenViewDiv().style.display = 'none';
+          MeetingManager.stopViewingScreenShare();
+          dispatch({ type: actionType.StopScreenShareView });
+          break;
+        default:
+          console.log(`Unhandled incoming message: ${type}`);
+          break;
+      }
+    } catch (e) {
+      alert(e);
+    }
+  };
+
+  useEffect(() => {
+    if (!isInitialized.current) return;
+
+    if (state.activeMeeting) {
+      sendMessage({ type: messageType.MeetingJoined });
+    } else {
+      sendMessage({ type: messageType.MeetingLeft });
+    }
+  }, [state.activeMeeting]);
+
+  useEffect(() => {
+    if (!isInitialized.current) return;
+
+    const { isSharingLocalVideo } = state;
+    sendMessage({
+      type: messageType.MeetingData,
+      payload: { isSharingLocalVideo },
+    });
+  }, [state.isSharingLocalVideo]);
+
+  useEffect(() => {
+    if (!isInitialized.current) return;
+
+    const { isViewingSharedScreen } = state;
+    sendMessage({
+      type: messageType.MeetingData,
+      payload: { isViewingSharedScreen },
+    });
+  }, [state.isViewingSharedScreen]);
+
+  useEffect(() => {
+    if (!window.deviceEnvironment) return;
+
+    window.deviceEnvironment.then(env => {
+      env.init(messageHandler);
+    });
+
+    isInitialized.current = true;
+  }, []);
+
+  return null;
+};
+
+export default withRouter(RoomProvider);
```

Comment:
optional (don't need to update): `react-router-dom` offers `useHistory`

Reply:
Updated!